File: problem57.py
class Solution:
    def insert(self, intervals: List[List[int]], newInterval: List[int]):

        # base case
        if len(intervals) == 0:
            return [newInterval]
        # Make new array with no overlapping
        answer = [] # new interval with new point
        change = True
        middle = False
        for i in intervals:
            if i[0] <= newInterval[0] <= i[1] and i[0] <= newInterval[1] <= i[1]:
                answer.append(i)    
                change = False
            # find the point where new start of interval exist 
            elif newInterval[0] <= i[0] <= newInterval[1] or newInterval[0] <= i[1] <= newInterval[1]: 
                if i[0] < newInterval[0]:
                    newInterval[0] = i[0] # update new interval start
                # find the exist point for new interval one
                if newInterval[1] < i[1]:
                    newInterval[1] = i[1] # update new interval end
                middle = True
            else:
                if middle:
                    answer.append(newInterval)
                    middle = False
                    change = False
                answer.append([i[0], i[1]]) # no need for any changes 
        
        if change:
            answer.append(newInterval)
        answer.sort()
        return answer

File: problem57.cpp
#include <vector>
#include <algorithm>
#include <iostream>
using namespace std;
class Solution {
public:
    vector<vector<int>> insert(vector<vector<int>>& intervals, vector<int>& newInterval) {
        // sorted --> add all intervals before newInterval 
        int i = 0; // index of intervals
        int len = intervals.size();
        vector<vector<int>> answer;

        // add all intervals before reaching newinterval starting point
        while(i < len && newInterval[0] > intervals[i][1]){
            answer.push_back(intervals[i]);
            i++;
        }

        // Now expand interval where overlap with newinterval
        while(i < len && intervals[i][0] <= newInterval[1]){
            newInterval[0] = min(intervals[i][0], newInterval[0]);
            newInterval[1] = max(intervals[i][1], newInterval[1]);
            i++;
        }
        answer.push_back(newInterval);

        // Now add remaining intervals

        while(i < len){
            answer.push_back(intervals[i]);
            i++;
        }

        return answer;
    }
};